src/core/Object3D.js

```diff
@@ -136,6 +205,8 @@ class Object3D extends EventDispatcher {
 
 		this.matrix.decompose( this.position, this.quaternion, this.scale );
 
+		this.matrixWorldNeedsUpdate = true;
+
```

Comment:
We relied on the world matrix being updated always, so we could update the local matrix through its methods, and expect the world matrix to update also. I'd prefer not to modify matrices or math classes, but that would be a sure way of handling this. Currently, if you mutate the matrix itself, you'll need to flip `matrixWorldNeedsUpdate` or handle that part yourself #28533.

Reply:
I'm wondering if the optimization where we no longer update the world matrix automatically unless position/scale/quaternion/rotation properties were mutated is too breaking (e.g. the local matrix is instead transiently updated with a mutation like in helpers, or via its methods). I would be okay downscoping this to just the local matrix, but that is less effective of an optimization for deeply nested scenes. Don't see a way of mitigating this without modifying matrices in some way.

```js
updateMatrix() {

	if ( this.matrixNeedsUpdate === true ) {

		this.matrix.compose( this.position, this.quaternion, this.scale );
		this.matrixNeedsUpdate = false;
		this.matrixWorldNeedsUpdate = true; // maybe should be set to true always?

	}

}
```

```js
function onLocalMatrixWrite() {

	self.matrixWorldNeedsUpdate = true;

}

listenToProperties( matrix.elements, Array.from( { length: 16 }, ( _, i ) => i ), undefined, onLocalMatrixWrite );
```